File: app/src/main/java/com/sjn/stamp/ui/item/RankedArtistItem.java
package com.sjn.stamp.ui.item;

import android.animation.Animator;
import android.app.Activity;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.sjn.stamp.R;
import com.sjn.stamp.utils.ViewHelper;

import java.io.Serializable;
import java.util.List;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.helpers.AnimatorHelper;
import eu.davidea.flexibleadapter.items.IFilterable;
import eu.davidea.flexibleadapter.utils.Utils;

public class RankedArtistItem extends AbstractItem<RankedArtistItem.SimpleViewHolder> implements IFilterable, Serializable {
    public int getOrder() {
        return mOrder;
    }

    public int getPlayCount() {
        return mPlayCount;
    }

    public String getArtistName() {
        return mArtistName;
    }

    private int mOrder;
    private int mPlayCount;
    private String mMostPlayedSongTitle;
    private String mArtistName;
    private String mArtUrl;
    private MediaMetadataCompat mTrack;

    public RankedArtistItem(MediaMetadataCompat track, String artistName, String artUrl, int playCount, int order) {
        super(artistName);
        setDraggable(true);
        setSwipeable(true);
        mTrack = track;
        mMostPlayedSongTitle = track.getDescription().getTitle().toString();
        mArtistName = artistName;
        mArtUrl = artUrl;
        mOrder = order;
        mPlayCount = playCount;
    }

    public MediaMetadataCompat getTrack() {
        return mTrack;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.recycler_ranked_item;
    }

    @Override
    public SimpleViewHolder createViewHolder(FlexibleAdapter adapter, LayoutInflater inflater, ViewGroup parent) {
        return new SimpleViewHolder(inflater.inflate(getLayoutRes(), parent, false), adapter);
    }

    @Override
    @SuppressWarnings({"unchecked"})
    public void bindViewHolder(final FlexibleAdapter adapter, SimpleViewHolder holder, int position, List payloads) {
        Context context = holder.itemView.getContext();
        if (adapter.hasSearchText()) {
            Utils.highlightText(holder.mTitle, getTitle(), adapter.getSearchText());
            Utils.highlightText(holder.mSubtitle, getSubtitle(), adapter.getSearchText());
        } else {
            holder.mTitle.setText(mArtistName);
            holder.mSubtitle.setText(context.getString(R.string.item_ranking_most_played, mMostPlayedSongTitle));
            holder.mCountView.setText(String.valueOf(mPlayCount));
            holder.mOrderView.setText(String.valueOf(mOrder));
        }
        if (mArtUrl != null) {
            ViewHelper.updateAlbumArt((Activity) context, holder.mAlbumArtView, mArtUrl, mArtistName);
        }
    }

    @Override
    public boolean filter(String constraint) {
        return mArtistName.toLowerCase().trim().contains(constraint);
    }

    static final class SimpleViewHolder extends LongClickDisableViewHolder {

        ImageView mAlbumArtView;
        TextView mTitle;
        TextView mSubtitle;
        Context mContext;
        View mFrontView;
        TextView mCountView;
        TextView mOrderView;

        SimpleViewHolder(View view, FlexibleAdapter adapter) {
            super(view, adapter);
            this.mContext = view.getContext();

            this.mTitle = view.findViewById(R.id.title);
            this.mSubtitle = view.findViewById(R.id.subtitle);
            this.mAlbumArtView = view.findViewById(R.id.image);
            this.mAlbumArtView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mAdapter.mItemLongClickListener != null) {
                        mAdapter.mItemLongClickListener.onItemLongClick(getAdapterPosition());
                        toggleActivation();
                    }
                }
            });

            this.mFrontView = view.findViewById(R.id.front_view);
            this.mCountView = view.findViewById(R.id.count);
            this.mOrderView = view.findViewById(R.id.order);
        }

        @Override
        public void toggleActivation() {
            super.toggleActivation();
        }

        @Override
        public float getActivationElevation() {
            return ViewHelper.dpToPx(itemView.getContext(), 4f);
        }

        @Override
        public View getFrontView() {
            return mFrontView;
        }

        @Override
        public void scrollAnimators(@NonNull List<Animator> animators, int position, boolean isForward) {
            if (mAdapter.getRecyclerView().getLayoutManager() instanceof GridLayoutManager ||
                    mAdapter.getRecyclerView().getLayoutManager() instanceof StaggeredGridLayoutManager) {
                if (position % 2 != 0)
                    AnimatorHelper.slideInFromRightAnimator(animators, itemView, mAdapter.getRecyclerView(), 0.5f);
                else
                    AnimatorHelper.slideInFromLeftAnimator(animators, itemView, mAdapter.getRecyclerView(), 0.5f);
            } else {
                //Linear layout
                if (mAdapter.isSelected(position))
                    AnimatorHelper.slideInFromRightAnimator(animators, itemView, mAdapter.getRecyclerView(), 0.5f);
                else
                    AnimatorHelper.slideInFromLeftAnimator(animators, itemView, mAdapter.getRecyclerView(), 0.5f);
            }
        }
    }

    @Override
    public String toString() {
        return "SongHistoryItem[" + super.toString() + "]";
    }

}

File: app/src/main/java/com/sjn/stamp/ui/item/RankedArtistItem.kt
package com.sjn.stamp.ui.item

import android.app.Activity
import android.content.Context
import android.support.v4.media.MediaMetadataCompat
import android.view.View
import com.sjn.stamp.R
import com.sjn.stamp.ui.item.holder.RankedViewHolder
import com.sjn.stamp.utils.ViewHelper
import eu.davidea.flexibleadapter.FlexibleAdapter
import eu.davidea.flexibleadapter.items.IFilterable
import eu.davidea.flexibleadapter.utils.FlexibleUtils
import java.io.Serializable

class RankedArtistItem(
        val context: Context,
        val track: MediaMetadataCompat,
        val artistName: String,
        artUrl: String?,
        val playCount: Int,
        val order: Int)
    : AbstractItem<RankedViewHolder>(artistName), IFilterable, Serializable {

    private val mostPlayedSongTitle = track.description.title?.toString() ?: ""
    override val title = artistName
    override val subtitle = context.getString(R.string.item_ranking_most_played, mostPlayedSongTitle)
            ?: ""
    private val albumArt = artUrl ?: ""

    init {
        isDraggable = true
        isSwipeable = true
    }

    override fun getLayoutRes(): Int = R.layout.recycler_ranked_item

    override fun createViewHolder(view: View, adapter: FlexibleAdapter<*>): RankedViewHolder = RankedViewHolder(view, adapter)

    override fun bindViewHolder(adapter: FlexibleAdapter<*>, holder: RankedViewHolder, position: Int, payloads: List<*>) {
        if (adapter.hasSearchText()) {
            FlexibleUtils.highlightText(holder.title, title, adapter.searchText)
            FlexibleUtils.highlightText(holder.subtitle, subtitle, adapter.searchText)
        } else {
            holder.title.text = title
            holder.subtitle.text = subtitle
            holder.countView.text = playCount.toString()
            holder.orderView.text = order.toString()
        }
        if (albumArt.isNotEmpty()) ViewHelper.updateAlbumArt(context as Activity, holder.albumArtView, albumArt, artistName)
    }

    override fun filter(constraint: String): Boolean = artistName.toLowerCase().trim { it <= ' ' }.contains(constraint)
}